Uaotilities/Delphi.py

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python
+# encoding: utf-8
+'''
+Delphi.py
+
+@author:     Piotr Nikiel <piotr.nikiel@gmail.com>
+@author:     Paris Moschovakos <paris.moschovakos@cern.ch>
+
+@copyright:  2020 CERN
+
+@license:
+Copyright (c) 2020, CERN.
+All rights reserved.
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer.
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT  HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS  OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT  LIMITED TO, THE IMPLIED
+WARRANTIES  OF  MERCHANTABILITY  AND  FITNESS  FOR  A  PARTICULAR  PURPOSE  ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL,  SPECIAL, EXEMPLARY, OR  CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT  NOT LIMITED TO,  PROCUREMENT OF  SUBSTITUTE GOODS OR  SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS  INTERRUPTION) HOWEVER CAUSED AND ON
+ANY  THEORY  OF  LIABILITY,   WHETHER IN  CONTRACT, STRICT  LIABILITY,  OR  TORT
+(INCLUDING  NEGLIGENCE OR OTHERWISE)  ARISING IN ANY WAY OUT OF  THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+@contact:    quasar-developers@cern.ch
+'''
+
+import sys
+
+sys.path.insert(0, 'FrameworkInternals')
+
+from Oracle import Oracle
+from transform_filters import cap_first
+
+Pythia = Oracle()
+
+class Delphi():
+
+    def readPronouncement(self, className, variableName, dataType):
+        output = dataType + " " + className + "::read" + cap_first(variableName) + " (\n"
```

Comment:
Since the read and write parts are common for source and cached variables I decided to "functionalize" instead of duplicating the code in the transform.
Probably I would have chosen a jinja2 macro here but I took your advice that this is strongly inferior to native implementation in Python. So I went this way.

Reply:
The main thing to avoid in Jinja transform is complex logic, especially if equivalent logic can be significantly simplified in Python.

However, looking at readPronouncement, I do not see any complex logic. In my viewpoint it is a perfect candidate to be put in Jinja, and I don't consider it a good candidate to be put in Python.

So I understand the problem is that it would be used twice, once from source-vars, once from cache-vars, in both cases still without any complex logic. Could we therefore consider either of:
(a) Jinja macro within that file, or
(b) an external Jinja template like e.g. here:
https://github.com/quasar-team/quasar/blob/master/Device/templates/designToRootHeader.jinja#L21

I think (a) would be preferred to (b) because the complexity is not that huge and the code would be shared by just one transform so no reason to need to go external. 

?

Reply:
I don't mind to redo this work with a different approach that is with macros, which would have been my initial preference exactly due to non-complexity. I do though have second thoughts on what is implemented already. Let me explain:

Questioning myself at this point, "what do we gain with the current implementation?" 

- We have a more sophisticated solution for what was probably needed which increase a bit the complexity but that doesn't go without some benefits.
- We uniformly have an equivalent to quasar mechanism to store "knowledge" which I think is a natural way for a quasar module symmetric to the its server "older" brother 
- That also allows for its nice expandability for the future where more knowledge will be need to be stored. We naturally know there is the place.
- We have a place to store more complex functions when needed. Since Uao is a lively project I think it will be needed anytime soon.
- And more importantly, the work is already done :)

On the other hand "what do we gain be discarding the above approach?"

- A visual simplification for the current Uao code generation as is today
- A more natural usage of Jinja2 abilities (this "output +=" repetition doesn't feel good... it doesn't at all)

Before deciding what to do here I just want to remind that I followed those instructions (concisely):

> notes for authors of quasar extension modules it should especially concern Paris (UaO translation)

> (5) In the old XSLT transforms there was a chance to define functions. Of course that was kind of clumsy. In Jinja you can achieve something similar with macros, that might be OK for simple cases but I expect this to be strongly inferior to native implementation in Python. So the question is: how do add custom functions to your new transforms without touching DesignInspector or Oracle?
> (a) we write the function(s) in the python program (i.e. the "main" program that generated things) of given extension module. Of course these functions can be also imported from secondary modules or even from 3rd party modules, no problem.

> examples of such implementation ... (def hilariousPythonFunction, additionalParam usage in transform design etc.)

Please take all the above into account and let me know what is your preference. I will just follow your decision ;)

Reply:
Hi,
I realize that the phrasing of the advice was perhaps slightly suboptimal.
My intent when writing the advice was to say, that if someone wants to do some complex logic/processing, one could add functions/methods (like you did in the Delphi) which would only return the minimum of processed information and AT THE SAME TIME keep most of the template in Jinja. When done this way, Jinja has most of template but none of complex logic, while Delphi has almost none of template but indeed does complex logic and processing.

In any case, as you also agreed in the writing, putting so long blocks of bare text as in readPronounciation is far from being easy to read and maintain. Let's talk offline (maybe during our meeting today) on how to improve it.